pkg/azure/aks/machine_store.go

```diff
@@ -0,0 +1,253 @@
+	m.machineMapLock.RLock()
+	defer m.machineMapLock.RUnlock()
+
+	vmInfo := m.MachineMap[vmName]
+	if vmInfo == nil {
+		return nil, ErrMachineNotFound
+	}
+
+	return vmInfo, nil
+}
+
+func (m *MachineStore) getVmInfoFromVmss(ctx context.Context, rgName string, vmssName string, priority MachinePriority, osInfo MachineOperatingSystem) (map[string]*VirtualMachineInfo, error) {
+	vmInfo := make(map[string]*VirtualMachineInfo)
+
+	opts := &armcompute.VirtualMachineScaleSetVMsClientListOptions{
+		Expand: to.StringPtr("instanceView"),
+	}
+	pager := m.azVMSSVmClient.NewListPager(rgName, vmssName, opts)
+	for pager.More() {
+		nextResult, err := pager.NextPage(ctx)
+		if err != nil {
+			m.logger.LogAttrs(ctx, slog.LevelError, "unable to advance page in VMSS VM Client", slog.String("err", err.Error()))
+			return nil, ErrPageAdvanceFailure
+		}
+
+		for _, v := range nextResult.Value {
+			vmName, err := getMachineName(v)
+			if err != nil {
+				m.logger.Error(err.Error())
+				continue
+			}
+
+			vmInfo[vmName] = &VirtualMachineInfo{
+				Name:            vmName,
+				Region:          to.String(v.Location),
+				OwningVMSS:      vmssName,
+				MachineTypeSku:  to.String(v.SKU.Name),
+				Priority:        priority,
+				OperatingSystem: osInfo,
+			}
+			m.logger.LogAttrs(ctx, slog.LevelDebug, "found machine information", slog.String("machineName", vmName))
+		}
+	}
+
+	return vmInfo, nil
+}
+
+func (m *MachineStore) getVmInfoFromResourceGroup(ctx context.Context, rgName, clusterName string) (map[string]*VirtualMachineInfo, error) {
+	vmInfoMap := make(map[string]*VirtualMachineInfo)
+
+	pager := m.azVMSSClient.NewListPager(rgName, nil)
+	for pager.More() {
+		nextResult, err := pager.NextPage(ctx)
+		if err != nil {
+			m.logger.LogAttrs(ctx, slog.LevelError, "unable to advance page in VMSS Client", slog.String("err", err.Error()))
+			return nil, ErrPageAdvanceFailure
+		}
+
+		for _, v := range nextResult.Value {
+			vmssName := to.String(v.Name)
+			vmssPriority := getMachineScaleSetPriority(v)
+			vmssOperationSystem := getMachineScaleSetOperatingSystem(v)
+
+			vmInfo, err := m.getVmInfoFromVmss(ctx, rgName, vmssName, vmssPriority, vmssOperationSystem)
+			if err != nil {
+				return nil, err
+			}
+
+			for name, info := range vmInfo {
+				info.OwningCluster = clusterName
+				vmInfoMap[name] = info
+			}
+		}
+	}
+
+	return vmInfoMap, nil
+}
+
+func (m *MachineStore) getClustersInSubscription(ctx context.Context) ([]*armcontainerservice.ManagedCluster, error) {
+	clusterList := []*armcontainerservice.ManagedCluster{}
+
+	pager := m.azAksClient.NewListPager(nil)
+	for pager.More() {
+		page, err := pager.NextPage(ctx)
+		if err != nil {
+			m.logger.LogAttrs(ctx, slog.LevelError, "unable to advance page in AKS Client", slog.String("err", err.Error()))
+			return nil, ErrPageAdvanceFailure
+		}
+		clusterList = append(clusterList, page.Value...)
+	}
+
+	return clusterList, nil
```

Comment:
Suggestion, to improve the ability to test this method __and__ make it reusable if we need to in the future, pass in the aks client as a parameter to the method and decouple the method from `MachineStore`.

Reply:
We've discussed this offline, but essentially the decision is:

- leave this as is for now
- work on mocking out the clients and then putting the mocked version on the object directly for tests rather than using the constructor